libraries/AP_Proximity/AP_Proximity_LightWareSerial.cpp

```diff
@@ -0,0 +1,153 @@
+#include <AP_Math/crc.h>
+#include "AP_Proximity_LightWareSerial.h"
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define PROXIMITY_LIGHTWARE_HEADER    0xAA
+
+// send message to sensor
+void AP_Proximity_LightWareSerial::send_message(uint8_t msgid, bool write, const uint8_t *payload, uint16_t payload_len)
+{
+    if ((_uart == nullptr) || (payload_len > PROXIMITY_LIGHTWARE_PAYLOAD_LEN_MAX)) {
+        return;
+    }
+
+    // check for sufficient space in outgoing buffer
+    if (_uart->txspace() < payload_len + 6U) {
+        return;
+    }
+
+    // write header
+    _uart->write((uint8_t)PROXIMITY_LIGHTWARE_HEADER);
+    uint16_t crc = crc_xmodem_update(0, PROXIMITY_LIGHTWARE_HEADER);
+
+    // write flags including payload length
+    const uint16_t flags = ((payload_len+1) << 6) | (write ? 0x01 : 0);
+    _uart->write(LOWBYTE(flags));
+    crc = crc_xmodem_update(crc, LOWBYTE(flags));
+    _uart->write(HIGHBYTE(flags));
+    crc = crc_xmodem_update(crc, HIGHBYTE(flags));
+
+    // msgid
+    _uart->write(msgid);
+    crc = crc_xmodem_update(crc, msgid);
+
+    // payload
+    if ((payload_len > 0) && (payload != nullptr)) {
+        for (uint16_t i = 0; i < payload_len; i++) {
+            _uart->write(payload[i]);
+            crc = crc_xmodem_update(crc, payload[i]);
+        }
+    }
+
+    // checksum
+    _uart->write(LOWBYTE(crc));
+    _uart->write(HIGHBYTE(crc));
+}
+
+// process one byte received on serial port
+// returns true if a complete message has been received
+// state is stored in _msg structure
+bool AP_Proximity_LightWareSerial::parse_byte(uint8_t b)
+{
+    // check that payload buffer is large enough
+    static_assert(ARRAY_SIZE(_msg.payload) == PROXIMITY_LIGHTWARE_PAYLOAD_LEN_MAX, "AP_Proximity_LightWareSerial: check _msg.payload array size");
+
+    // process byte depending upon current state
+    switch (_msg.state) {
+
+    case ParseState::HEADER:
+        if (b == PROXIMITY_LIGHTWARE_HEADER) {
+            _msg.crc_expected = crc_xmodem_update(0, b);
+            _msg.state = ParseState::FLAGS_L;
+        }
+        break;
+
+    case ParseState::FLAGS_L:
+        _msg.flags_low = b;
+        _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        _msg.state = ParseState::FLAGS_H;
+        break;
+
+    case ParseState::FLAGS_H:
+        _msg.flags_high = b;
+        _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        _msg.payload_len = UINT16_VALUE(_msg.flags_high, _msg.flags_low) >> 6;
+        if ((_msg.payload_len == 0) || (_msg.payload_len > PROXIMITY_LIGHTWARE_PAYLOAD_LEN_MAX)) {
+            // invalid payload length, abandon message
+            _msg.state = ParseState::HEADER;
+        } else {
+            _msg.state = ParseState::MSG_ID;
+        }
+        break;
+
+    case ParseState::MSG_ID:
+        _msg.msgid = b;
+        _msg.crc_expected = crc_xmodem_update(_msg.crc_expected, b);
+        if (_msg.payload_len > 1) {
```

Comment:
Not zero?

Reply:
aha! not, it's not a mistake actually.  It's very odd but the message payload includes the message id (see here: http://support.lightware.co.za/sf45/#/serial?id=packets).  It's weird and I brought it up with Lightware back when we added support for the LW20 which uses the same packet structure.